com.musala.simple.students.db/src/main/java/com/musala/simple/students/db/DatabaseFactory.java

```diff
@@ -0,0 +1,189 @@
+package com.musala.simple.students.db;
+
+import java.util.Properties;
+
+import com.musala.simple.students.db.DatabaseTypes;
+
+/**
+ * This is a Factory class for creating {@link Database} objects. The
+ * constructor requires the database properties to be provided in the form of a
+ * {@link Properties} object passed to the constructor along with the database
+ * type {@link DatabaseTypes} enum.
+ * 
+ * 
+ * @author yoan.petrushinov
+ *
+ */
+public class DatabaseFactory {
+	private static final String DEFAULT_DB_NAME = "studentsDb";
+	private static final int DEFAULT_DB_PORT = 27017;
+	private static final String DEFAULT_DB_HOST = "localhost";
+	private static final String DEFAULT_DB_USERNAME = "admin";
+	private static final String DEFAULT_DB_PASSWORD = "admin";
+
+	private static final String DB_NAME = "name";
+	private static final String DB_PORT = "port";
+	private static final String DB_HOST = "host";
+	private static final String DB_USER = "username";
+	private static final String DB_PASSWORD = "password";
+	private DatabaseTypes dbType;
+	private String name;
+	private int port;
+	private String host;
+	private String username;
+	private String password;
+
+	public DatabaseFactory(Properties dbProperties, DatabaseTypes dbType) {
+		this.setDbType(dbType);
+		this.setName(dbProperties.getProperty(DB_NAME));
+		this.setPort(dbProperties.getProperty(DB_PORT));
+		this.setHost(dbProperties.getProperty(DB_HOST));
+		this.setUsername(dbProperties.getProperty(DB_USER));
+		this.setPassword(dbProperties.getProperty(DB_PASSWORD));
+	}
+
+	/**
+	 * Creates a new implementation of the {@link Database}.
+	 * 
+	 * @return a new database of the required type
+	 */
+	public Database getDatabase() {
```

Comment:
The factory is fine, but we can make it better! :) Let's change this method to be named `public static Database createDatabase(DatabaseTypes dbType)` and implement the so-called Builder pattern (often supported by factories). Perform the following steps:

- Remove the implemented constructor, we won't need it.
- The static method `createDatabase(DatabaseTypes dbType)` (previously `getDatabase()`) should internally do the currently implemented logic with the switch statement and return the newly created `Database` object (MySQL database, Mondo database, etc).
- Then, all setters should be renamed and re-declared too - e.g. `setName()` has to become `withName(String name)` and internally the logic should be - we call `withName(dbProperties.getProperty(DB_NAME))` and underneath it does `this.name = name` and then `return this`.
- Keep in mind that making `createDatabase()` static will require all with-something methods to be static.
- Finally, you will have to have a `public Database build()` method which returns `this` instance.
- The getter methods, on the other hand, can be removed - they don't seem to be used anywhere.

The whole point is that, externally, I can do the following call:

```java
Properties dbProperties = FileHelper.readDbPropertiesFile(DatabaseTypes.MongoDb);
Database database = DatabaseFactory.createDatabase(DatabaseTypes.MongoDb).withName(dbProperties.getProperty(DB_NAME))withHost(dbProperties.getProperty(DB_HOST)).withPort(...).withUsername(...).withPassword().build().
```

You can use this github class as a reference: https://github.com/eclipse/smarthome/blob/4204ce06bb28c28e5f711e720f87ef83beff2e27/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/builder/ThingBuilder.java